File: ntpclock/src/internal/clock_corrector.cc
// Copyright (c) 2025 The NTP Sample Authors
#include "internal/clock_corrector.hpp"

#include <algorithm>
#include <cmath>

namespace ntpclock {
namespace internal {

Status::Correction ClockCorrector::Apply(
    double current_offset, double target_offset, double step_threshold_s,
    double slew_rate_s_per_s, double poll_interval_s, double* out_amount_s) {
  double delta = target_offset - current_offset;

  if (std::abs(delta) >= step_threshold_s) {
    // Step correction: apply full delta immediately
    ntpserver::TimeSpec new_offset =
        ntpserver::TimeSpec::FromDouble(current_offset + delta);
    ntpserver::TimeSpec delta_ts = ntpserver::TimeSpec::FromDouble(delta);
    {
      std::lock_guard<std::mutex> lk(mtx_);
      offset_applied_ = new_offset;
      // Adjust last_returned_ by the step amount to allow backward jump
      last_returned_ = last_returned_ + delta_ts;
      allow_backward_once_ = true;
    }
    *out_amount_s = delta;
    return Status::Correction::Step;

  } else {
    // Slew correction: apply bounded-rate change
    double max_change = slew_rate_s_per_s * poll_interval_s;
    double change = std::clamp(delta, -max_change, max_change);

    if (std::abs(change) > 0.0) {
      ntpserver::TimeSpec new_offset =
          ntpserver::TimeSpec::FromDouble(current_offset + change);
      {
        std::lock_guard<std::mutex> lk(mtx_);
        offset_applied_ = new_offset;
      }
      *out_amount_s = change;
      return Status::Correction::Slew;
    }
  }

  *out_amount_s = 0.0;
  return Status::Correction::None;
}

void ClockCorrector::AllowBackwardOnce() {
  std::lock_guard<std::mutex> lk(mtx_);
  allow_backward_once_ = true;
}

ntpserver::TimeSpec ClockCorrector::GetMonotonicTime(
    const ntpserver::TimeSpec& base_time) {
  std::lock_guard<std::mutex> lk(mtx_);

  // Apply current offset correction
  ntpserver::TimeSpec candidate = base_time + offset_applied_;

  // Allow one backward jump right after a step correction
  if (allow_backward_once_) {
    allow_backward_once_ = false;
    last_returned_ = candidate;
    return candidate;
  }

  // Otherwise, enforce monotonicity with a small epsilon
  ntpserver::TimeSpec eps = ntpserver::TimeSpec::FromDouble(1e-9);
  ntpserver::TimeSpec min_next = last_returned_ + eps;

  if (candidate < min_next) {
    last_returned_ = min_next;
    return min_next;
  } else {
    last_returned_ = candidate;
    return candidate;
  }
}

double ClockCorrector::GetOffsetApplied() const {
  std::lock_guard<std::mutex> lk(mtx_);
  return offset_applied_.ToDouble();
}

void ClockCorrector::ResetOffset() {
  std::lock_guard<std::mutex> lk(mtx_);
  offset_applied_ = ntpserver::TimeSpec{};
  last_returned_ = ntpserver::TimeSpec{};
  allow_backward_once_ = false;
}

}  // namespace internal
}  // namespace ntpclock

File: csharp/NtpClock/Internal/ClockCorrector.cs
// Copyright (c) 2025 <Your Name>
using NtpServer.Internal;

namespace NtpClock.Internal;

/// <summary>
/// Unified clock correction and monotonicity management.
/// Combines correction decision logic (slew vs step) with monotonic time
/// enforcement. When a step correction is applied, coordinates the one-time
/// backward jump allowance for NowUnix().
/// </summary>
public class ClockCorrector
{
    private readonly object _lock = new object();
    private TimeSpec _offsetApplied = new TimeSpec();
    private TimeSpec _lastReturned = new TimeSpec();
    private bool _allowBackwardOnce = false;

    /// <summary>
    /// Apply offset correction based on target vs current delta.
    /// Called periodically from the sync loop to update the offset correction
    /// parameter. Does NOT read time; only updates the correction state.
    /// </summary>
    /// <param name="currentOffset">Currently applied offset (seconds).</param>
    /// <param name="targetOffset">Desired target offset (seconds).</param>
    /// <param name="stepThresholdS">Threshold for step correction (seconds).</param>
    /// <param name="slewRateSPerS">Maximum slew rate (seconds per second).</param>
    /// <param name="pollIntervalS">Polling interval (seconds).</param>
    /// <param name="outAmountS">Output parameter for actual correction amount applied.</param>
    /// <returns>Type of correction applied (None, Slew, or Step).</returns>
    public Status.CorrectionType Apply(
        double currentOffset,
        double targetOffset,
        double stepThresholdS,
        double slewRateSPerS,
        double pollIntervalS,
        out double outAmountS
    )
    {
        double delta = targetOffset - currentOffset;

        if (Math.Abs(delta) >= stepThresholdS)
        {
            // Step correction: apply full delta immediately
            TimeSpec newOffset = TimeSpec.FromDouble(currentOffset + delta);
            TimeSpec deltaTs = TimeSpec.FromDouble(delta);

            lock (_lock)
            {
                _offsetApplied = newOffset;
                // Adjust last_returned by the step amount to allow backward jump
                _lastReturned = _lastReturned + deltaTs;
                _allowBackwardOnce = true;
            }

            outAmountS = delta;
            return Status.CorrectionType.Step;
        }
        else
        {
            // Slew correction: apply bounded-rate change
            double maxChange = slewRateSPerS * pollIntervalS;
            double change = Math.Clamp(delta, -maxChange, maxChange);

            if (Math.Abs(change) > 0.0)
            {
                TimeSpec newOffset = TimeSpec.FromDouble(currentOffset + change);

                lock (_lock)
                {
                    _offsetApplied = newOffset;
                }

                outAmountS = change;
                return Status.CorrectionType.Slew;
            }
        }

        outAmountS = 0.0;
        return Status.CorrectionType.None;
    }

    /// <summary>
    /// Allow the next time reading to go backward once.
    /// Typically called after external time adjustments (e.g., vendor hints
    /// applying SetAbsolute) to permit GetMonotonicTime() to reflect the backward
    /// jump.
    /// </summary>
    public void AllowBackwardOnce()
    {
        lock (_lock)
        {
            _allowBackwardOnce = true;
        }
    }

    /// <summary>
    /// Get monotonic corrected time.
    /// Called from NowUnix() at arbitrary times to read the current corrected
    /// time. Applies the offset correction and enforces monotonic progression.
    /// </summary>
    /// <param name="baseTime">The raw base time from TimeSource.</param>
    /// <returns>The monotonic corrected time (base + offset, with monotonicity
    /// enforced, unless backward jump is explicitly allowed).</returns>
    public TimeSpec GetMonotonicTime(TimeSpec baseTime)
    {
        lock (_lock)
        {
            // Apply current offset correction
            TimeSpec candidate = baseTime + _offsetApplied;

            // Allow one backward jump right after a step correction
            if (_allowBackwardOnce)
            {
                _allowBackwardOnce = false;
                _lastReturned = candidate;
                return candidate;
            }

            // Otherwise, enforce monotonicity with a small epsilon
            TimeSpec eps = TimeSpec.FromDouble(1e-9);
            TimeSpec minNext = _lastReturned + eps;

            if (candidate < minNext)
            {
                _lastReturned = minNext;
                return minNext;
            }
            else
            {
                _lastReturned = candidate;
                return candidate;
            }
        }
    }

    /// <summary>
    /// Get current offset correction value.
    /// </summary>
    /// <returns>Current applied offset (in seconds as double for compatibility).</returns>
    public double GetOffsetApplied()
    {
        lock (_lock)
        {
            return _offsetApplied.ToDouble();
        }
    }

    /// <summary>
    /// Reset offset to zero.
    /// </summary>
    public void ResetOffset()
    {
        lock (_lock)
        {
            _offsetApplied = new TimeSpec();
            _lastReturned = new TimeSpec();
            _allowBackwardOnce = false;
        }
    }
}
